src/syntax/concrete.rs

```diff
@@ -0,0 +1,58 @@
+use crate::syntax::token::{SourceRange, TokenKind};
+
+pub trait DeclVisitor {
+    type DeclResult;
+
+    fn visit_var_decl(&mut self, var_decl: &VarDecl) -> Self::DeclResult;
+}
+
+pub trait ExprVisitor {
+    type ExprResult;
+}
+
+pub trait TypeVisitor {
+    type TypeResult;
+}
+
+pub trait StmtVisitor {
+    type StmtResult;
+}
+
+pub trait ConcreteVisitor : DeclVisitor + ExprVisitor + TypeVisitor + StmtVisitor {
```

Comment:
@HoshinoTented Do you like such Visitor API? Or you prefer something else?

Reply:
Ummmm, what is Visitor?

Reply:
Finally I gave up visitor pattern and I finally choose to use `match`.